File: backend-layer/src/main/java/io/redis/devrel/workshop/services/FilesProcessor.java
package io.redis.devrel.workshop.services;

import dev.langchain4j.data.document.Document;
import dev.langchain4j.data.document.DocumentParser;
import dev.langchain4j.data.document.DocumentSplitter;
import dev.langchain4j.data.document.parser.apache.pdfbox.ApachePdfBoxDocumentParser;
import dev.langchain4j.data.document.splitter.DocumentByParagraphSplitter;
import dev.langchain4j.data.segment.TextSegment;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

import java.io.File;
import java.io.FileInputStream;
import java.io.InputStream;
import java.util.List;

@Service
public class FilesProcessor {

    private final Logger logger = LoggerFactory.getLogger(FilesProcessor.class);

    @Autowired
    private MemoryService memoryService;

    @Value("${knowledge.base.input.files}")
    private String knowledgeBaseInputFiles;

    @Scheduled(fixedRate = 5000)
    public void scanForPdfFiles() {
        File dir = new File(knowledgeBaseInputFiles);
        if (dir.exists() && dir.isDirectory()) {
            File[] pdfFiles = dir.listFiles((d, name) -> name.toLowerCase().endsWith(".pdf"));
            if (pdfFiles != null) {
                for (File pdf : pdfFiles) {
                    processFile(pdf);
                }
            }
        }
    }

    private void processFile(File file) {
        final DocumentParser documentParser = new ApachePdfBoxDocumentParser(true);
        final DocumentSplitter documentSplitter = new DocumentByParagraphSplitter(1000, 100);

        logger.info("Processing file {}", file.getAbsolutePath());
        try (InputStream inputStream = new FileInputStream(file)) {
            Document document = documentParser.parse(inputStream);
            if (document == null || document.text() == null || document.text().isBlank()) {
                logger.warn("Empty or unparseable document: {}", file.getName());
                renameToProcessed(file);
                return;
            }

            List<TextSegment> segments = documentSplitter.split(document);
            if (segments.isEmpty()) {
                logger.warn("No segments created from document: {}", file.getName());
                renameToProcessed(file);
                return;
            }

            int chunksStored = 0;
            for (int i = 0; i < segments.size(); i++) {
                TextSegment segment = segments.get(i);
                if (segment.text().trim().length() < 50) {
                    continue;
                }
                String entryText = formatSegmentWithMetadata(
                        segment.text(),
                        file.getName(),
                        i + 1,
                        segments.size()
                );
                try {
                    memoryService.createKnowledgeBaseEntry(entryText);
                    chunksStored++;
                } catch (Exception e) {
                    logger.warn("Failed to store segment {} of {} from {}", i + 1, segments.size(), file.getName(), e);
                }
            }
            logger.info("Processed {} - {} segments stored out of {} total", file.getName(), chunksStored, segments.size());
        } catch (Exception e) {
            logger.error("Failed to process {}", file.getName(), e);
        } finally {
            renameToProcessed(file);
        }
    }

    private void renameToProcessed(File file) {
        String filePath = file.getAbsolutePath();
        if (filePath.toLowerCase().endsWith(".pdf")) {
            File processedFile = new File(filePath.substring(0, filePath.length() - 4) + ".processed");
            boolean renamed = file.renameTo(processedFile);
            if (!renamed) {
                logger.warn("Could not rename file {} to .processed", file.getName());
            } else {
                logger.info("Renamed file {} to {}", file.getName(), processedFile.getName());
            }
        }
    }

    private String formatSegmentWithMetadata(String segmentText, String fileName, int segmentNumber, int totalSegments) {
        String metadata = String.format("[Document: %s | Section %d of %d]\n", fileName, segmentNumber, totalSegments);
        return metadata + segmentText.trim();
    }
}

File: backend-dotnet-layer/Services/FilesProcessor.cs
using System.Text;
using Microsoft.Extensions.Options;
using UglyToad.PdfPig;
using UglyToad.PdfPig.Content;
using UglyToad.PdfPig.DocumentLayoutAnalysis.TextExtractor;
using BackendDotnetLayer.Configuration;

namespace BackendDotnetLayer.Services;

public sealed class FilesProcessor : BackgroundService
{
    private readonly FilesProcessorOptions _options;
    private readonly ILogger<FilesProcessor> _logger;
    private readonly IServiceScopeFactory _serviceScopeFactory;

    public FilesProcessor(
        IOptions<FilesProcessorOptions> options,
        ILogger<FilesProcessor> logger,
        IServiceScopeFactory serviceScopeFactory)
    {
        _options = options.Value;
        _logger = logger;
        _serviceScopeFactory = serviceScopeFactory;
    }

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        var interval = TimeSpan.FromSeconds(Math.Max(1, _options.ScanIntervalSeconds));

        while (!stoppingToken.IsCancellationRequested)
        {
            try
            {
                await ScanForPdfFilesAsync(stoppingToken);
            }
            catch (OperationCanceledException) when (stoppingToken.IsCancellationRequested)
            {
                break;
            }
            catch (Exception exception)
            {
                _logger.LogError(exception, "Unhandled error while scanning for PDF files");
            }

            await Task.Delay(interval, stoppingToken);
        }
    }

    private async Task ScanForPdfFilesAsync(CancellationToken cancellationToken)
    {
        var inputDirectory = ResolveInputDirectory();
        if (string.IsNullOrWhiteSpace(inputDirectory) || !Directory.Exists(inputDirectory))
        {
            return;
        }

        foreach (var pdfPath in Directory.EnumerateFiles(inputDirectory, "*.pdf", SearchOption.TopDirectoryOnly))
        {
            await ProcessFileAsync(pdfPath, cancellationToken);
        }
    }

    private async Task ProcessFileAsync(string filePath, CancellationToken cancellationToken)
    {
        _logger.LogInformation("Processing file {FilePath}", filePath);

        try
        {
            var documentText = ExtractText(filePath);
            if (string.IsNullOrWhiteSpace(documentText))
            {
                _logger.LogWarning("Empty or unparseable document: {FileName}", Path.GetFileName(filePath));
                return;
            }

            var segments = SplitIntoSegments(documentText);
            if (segments.Count == 0)
            {
                _logger.LogWarning("No segments created from document: {FileName}", Path.GetFileName(filePath));
                return;
            }

            var chunksStored = 0;
            using var scope = _serviceScopeFactory.CreateScope();
            var memoryService = scope.ServiceProvider.GetRequiredService<MemoryService>();

            for (var index = 0; index < segments.Count; index++)
            {
                cancellationToken.ThrowIfCancellationRequested();

                var segment = segments[index].Trim();
                if (segment.Length < _options.MinimumChunkLength)
                {
                    continue;
                }

                var entryText = FormatSegmentWithMetadata(segment, Path.GetFileName(filePath), index + 1, segments.Count);
                try
                {
                    await memoryService.CreateKnowledgeBaseEntryAsync(entryText, cancellationToken);
                    chunksStored++;
                }
                catch (Exception exception)
                {
                    _logger.LogWarning(exception, "Failed to store segment {SegmentNumber} of {TotalSegments} from {FileName}", index + 1, segments.Count, Path.GetFileName(filePath));
                }
            }

            _logger.LogInformation("Processed {FileName} - {ChunksStored} segments stored out of {TotalSegments} total", Path.GetFileName(filePath), chunksStored, segments.Count);
        }
        catch (Exception exception)
        {
            _logger.LogError(exception, "Failed to process {FileName}", Path.GetFileName(filePath));
        }
        finally
        {
            RenameToProcessed(filePath);
        }
    }

    private string ExtractText(string filePath)
    {
        using var document = PdfDocument.Open(filePath);
        var builder = new StringBuilder();

        foreach (var page in document.GetPages())
        {
            var pageText = ContentOrderTextExtractor.GetText(page);
            if (string.IsNullOrWhiteSpace(pageText))
            {
                continue;
            }

            if (builder.Length > 0)
            {
                builder.AppendLine();
                builder.AppendLine();
            }

            builder.Append(pageText.Trim());
        }

        return builder.ToString();
    }

    private List<string> SplitIntoSegments(string documentText)
    {
        var paragraphs = documentText
            .Split(["\r\n\r\n", "\n\n"], StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
            .Where(paragraph => !string.IsNullOrWhiteSpace(paragraph))
            .ToList();

        if (paragraphs.Count == 0)
        {
            return [];
        }

        var segments = new List<string>();
        var current = new StringBuilder();

        foreach (var paragraph in paragraphs)
        {
            var normalizedParagraph = paragraph.Replace("\r", " ").Replace("\n", " ").Trim();
            if (string.IsNullOrWhiteSpace(normalizedParagraph))
            {
                continue;
            }

            if (normalizedParagraph.Length > _options.MaxChunkLength)
            {
                FlushCurrentSegment(segments, current);
                segments.AddRange(SplitLongParagraph(normalizedParagraph));
                continue;
            }

            var separatorLength = current.Length == 0 ? 0 : 2;
            if (current.Length + separatorLength + normalizedParagraph.Length > _options.MaxChunkLength)
            {
                FlushCurrentSegment(segments, current);
            }

            if (current.Length > 0)
            {
                current.AppendLine();
                current.AppendLine();
            }

            current.Append(normalizedParagraph);
        }

        FlushCurrentSegment(segments, current);
        return segments;
    }

    private List<string> SplitLongParagraph(string paragraph)
    {
        var segments = new List<string>();
        var start = 0;
        var overlap = Math.Max(0, Math.Min(_options.ChunkOverlapLength, _options.MaxChunkLength / 2));

        while (start < paragraph.Length)
        {
            var length = Math.Min(_options.MaxChunkLength, paragraph.Length - start);
            var slice = paragraph.Substring(start, length).Trim();
            if (!string.IsNullOrWhiteSpace(slice))
            {
                segments.Add(slice);
            }

            if (start + length >= paragraph.Length)
            {
                break;
            }

            start += Math.Max(1, length - overlap);
        }

        return segments;
    }

    private static void FlushCurrentSegment(ICollection<string> segments, StringBuilder current)
    {
        if (current.Length == 0)
        {
            return;
        }

        segments.Add(current.ToString().Trim());
        current.Clear();
    }

    private static string FormatSegmentWithMetadata(string segmentText, string fileName, int segmentNumber, int totalSegments)
    {
        return $"[Document: {fileName} | Section {segmentNumber} of {totalSegments}]{Environment.NewLine}{segmentText.Trim()}";
    }

    private void RenameToProcessed(string filePath)
    {
        if (!filePath.EndsWith(".pdf", StringComparison.OrdinalIgnoreCase))
        {
            return;
        }

        var processedPath = Path.ChangeExtension(filePath, ".processed");
        try
        {
            if (File.Exists(processedPath))
            {
                File.Delete(processedPath);
            }

            File.Move(filePath, processedPath);
            _logger.LogInformation("Renamed file {OriginalName} to {ProcessedName}", Path.GetFileName(filePath), Path.GetFileName(processedPath));
        }
        catch (Exception exception)
        {
            _logger.LogWarning(exception, "Could not rename file {FileName} to .processed", Path.GetFileName(filePath));
        }
    }

    private string ResolveInputDirectory()
    {
        return Environment.GetEnvironmentVariable("KNOWLEDGE_BASE_INPUT_FILES")
            ?? _options.InputDirectory;
    }
}
